pages/posts/the-architectural-scope-of-json-schema.md

```diff
@@ -0,0 +1,76 @@
+
+The reasons to ensure we have valid JSON data could be numerous: checking data before deserializing into programmatic models, checking form inputs before submission, etc.  **These are the needs of applications.**
+
+So _applications_ are the consumers of JSON Schema.
+
+But an application can't consume a specification without that specification being realized into code.  That's where implementations come in.
+
+An implementation of JSON Schema is an embodiment of the specification that is directly consumable by an application.
+
+## A nuance
+
+Something I've seen a lot of lately, and I think the source of some of the confusion that has arisen in my discussions, is that an application may be merely an executable wrapper around an implementation.  This tends to give the appearance that the application itself is the implementation, in which case the application would be subject to the requirements of the specification.  But I don't see it that way.  Even in these cases, there exists a distinction between the application and the implementation, even if that distinction is, in practice, really blurry.
+
+Applications tend to have three basic components:  an interface (UX or API), some business logic, and data persistence.  **All** applications have an interface.  However, the business logic and data persistence components are optional to the degree that you can have one or the other or both.  (An application with only a UX is generally not very useful.)
+
+An application may only provide an interface over data persistence (e.g. a Postgres web service), meaning that there's no need for any business logic.  Conversely, another application may provide a computational service (e.g. image processing) where there's no need to persist data.
+
+For the recent conversations I've had, this second scenario seems to be the case: an application is created that just evaluates instances against schemas.  But this doesn't mean that the application and implementation *are* the same thing.
+
+Within these applications the implementation, which for these applications _is_ the business logic, is still a separate component from the interface.  And it's important to recognize that JSON Schema as a specification can only cover the implementation part.
+
+## Why any of this matters
+
+It all comes back to what I touched on this in the opening section: JSON Schema needs to define inputs and outputs.  This comprises a minimal API that implementations and applications can use to communicate with each other.
+
+When the line between implementation and application is blurred, it's natural to think that the specification is imposing requirements on how the application communicates with its users.  But that's not the case.
+
+It is impossible for JSON Schema to know the needs of an application's users, and so it's impractical for the specification to attempt to define input and output requirements to which applications must adhere.
```

Comment:
I think if you explicitly contrast applications with libraries here this will be more clear.  "define input and output requirements to which applications must adhere" could be talking about either the application<->library interface (which _is_ what JSON Schema defines) or the user<->application interface (which JSON Schema cannot and should not define).

Reply:
Again, the implementation _may not_ be a library.  It _could_ be built in to the application (e.g. reimplemented instead of using an existing lib to cater for some special need of the application).  I want to isolate the JSON Schema logic, though, as something the spec has say on.